gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,482 @@
+        if self.waiting_step or (not self.waiting_reset):
+            logger.error('Calling `reset_wait` without any prior call to '
+                '`reset_async`. Closing `{0}`...'.format(type(self).__name__))
+            self.close(terminate=True)
+            raise NotResettingError('Calling `reset_wait` without any prior '
+                'call to `reset_async`.')
+
+        if not self.poll(timeout):
+            self.waiting_reset = False
+            raise mp.TimeoutError('The call to `reset_wait` has timed out after '
+                '{0} second{1}.'.format(timeout, 's' if timeout > 1 else ''))
+
+        self._restart_if_errors(reset=True)
+        observations_list = [pipe.recv() for pipe in self.parent_pipes]
+        self.waiting_reset = False
+
+        if not self.shared_memory:
+            concatenate(observations_list, self.observations,
+                self.single_observation_space)
+
+        return deepcopy(self.observations) if self.copy else self.observations
+
+    def step_async(self, actions):
+        """
+        Parameters
+        ----------
+        actions : iterable of samples from `action_space`
+            List of actions.
+        """
+        self._assert_is_running()
+        if self.waiting_reset:
+            logger.warn('Calling `step_async` while waiting for a pending '
+                'call to `reset` to complete. Waiting for the previous call '
+                'to `reset`...')
+            try:
+                self.reset_wait()
+            except mp.TimeoutError:
+                pass
+
+        if self.waiting_step:
+            logger.error('Calling `step_async` while waiting for a pending '
+                'call to `step` to complete. Closing `{0}`...'.format(
+                type(self).__name__))
+            self.close(terminate=True)
+            raise AlreadySteppingError('Calling `step_async` while waiting for '
+                'a pending call to `step` to complete.')
+
+        for pipe, action in zip(self.parent_pipes, actions):
+            pipe.send(('step', action))
+        self.waiting_step = True
+
+    def step_wait(self, timeout=None):
+        """
+        Parameters
+        ----------
+        timeout : int, optional
+            Number of seconds before the call to `step_wait` times out. If
+            `None`, the call to `step_wait` never times out.
+
+        Returns
+        -------
+        observations : sample from `observation_space`
+            A batch of observations from the vectorized environment.
+
+        rewards : `np.ndarray` instance (dtype `np.float_`)
+            A vector of rewards from the vectorized environment.
+
+        dones : `np.ndarray` instance (dtype `np.bool_`)
+            A vector whose entries indicate whether the episode has ended.
+
+        infos : list of dict
+            A list of auxiliary diagnostic informations.
+        """
+        self._assert_is_running()
+        if self.waiting_reset or not self.waiting_step:
```

Comment:
rather than dealing with a bunch of booleans, you can use a single state variable like this: https://github.com/cshesse/libenv/blob/master/libenv/libenv.py#L102